### PR TITLE
Ship to

### PR DESCRIPTION
**Pull request by bvoskr**

### CODE REVIEW

**Review by elderling — CHANGES_REQUESTED** [review]
I don't see the value added by the ['SetOnce'] trait, but if we're going to use it here then we need explain why some attributes have it and others do not. Some attributes have aliases and others do not but the aliases don't match the .xsd and the documentation in the class doesn't justify the need for them. If the aliases are justified for some reason I don't understand, the attributes with aliases don't have predicates for those aliases.

**Comment by bvoskr** [comment]
As for aliases, I simply saw them as a way to simplify request initialization for the developer. E.g., allow to use 'building' instead of 'buildingNumber', 'zip' instead of 'postal_code', 'phone' instead of 'phone_number' etc. Didn't put any more meaning to this. Since the only usage for alias was request initialization, I didn't care for predicates for aliases, too.

As for SetOnce trait, there are two topics which I want to separate:
1. 'ro' vs 'rw' attributes
2. Applying SetOnce trait to 'rw' attributes.

### 'ro' vs 'rw' attributes. 
As I see it (which is often wrong), 'rw' attributes are the ones that allowed to be added/modified in existing request for the convenience purposes (like adding building number, or phone number, etc. ) They do not redefine the delivery recipient, and do not affect funds calculation.
The attribute may be non-required, but ‘ro’ - e.g. first/last name, as it identifies the recipient.
The tests now are setting attributes during ShipTo initialization, and will probably pass even if I make all attributes 'ro'. But I’m quite sure division of attributes into 'ro' and 'rw' must remain.

### Applying SetOnce trait to 'rw' attributes.
Whether to apply SetOnce trait to the 'rw' attributes, I really do not have the clear answer. I was following pattern from “BillTo” field, which was applying SetOnce trait to all ‘rw’ fields. I'm ready to remove the trait.

**Review by elderling — CHANGES_REQUESTED** [review]

File: lib/MooseX/Types/CyberSource.pm [hunk 1]
```diff
@@ -278,6 +290,14 @@ subtype _VarcharSixty, as NonEmptySimpleStr,
   where { length $_ <= 60 },
   message { $varchar_message . '60' };
 
+subtype ShippingMethod, as _VarcharTen,
```
Comment (bvoskr) [inline lib/MooseX/Types/CyberSource.pm]: I actually thought about this. But SO API specs mentions two separate requirements for shipping method. One is, that max field length is 10, and another, listing possible supported values (none of them reaching max field length).
Enum would only cover supported values requirement. Current implementation covers both.

File: t/shipto.t [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    buildingNumber => 'N/A',
+    city           => 'Austin',
+    state          => 'TX',
+    country        => 'US',
+    postalCode     => '78701',
+);
+
+method_ok $ship_to, serialize => [], \%expected_serialized;
+
+# commenting code below till understand how MooseX works with undefined attributes
```
Comment (bvoskr) [inline t/shipto.t]: Will remove.

File: t/shipto-exceptions.t [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+use strict;
+use warnings;
+use Test::More;
+use Test::Fatal qw(exception dies_ok);
+
+use Module::Runtime 'use_module';
+
+my $shipto_c = use_module('Business::CyberSource::RequestPart::ShipTo');
+
```
Comment (bvoskr) [inline t/shipto-exceptions.t]: I'm missing the coverage context. It is either about sandbox being able to swallow all the fields, or about the integrity of the ShipTo object. 

- If former, I agree, and add all possible fields. It is, though, related to xt/ tests, not to t/shipto-exceptions.t

- If latter, t/shipto-exceptions.t seem to cover all the fail cases. t/shipto.t must be extended with missing fields.

**Review by elderling — APPROVED** [review]

**Pull request merged**